include/zephyr/async.h

```diff
@@ -0,0 +1,107 @@
+/*
```

Comment:
yeah perhaps, one of the issues with the proposal from @tbursztyka was the rom size increase because certain code paths couldn't be gc'ed as I read it.

I'll take a look at sys_notify. Really I wanted to avoid a single struct with a type tagged union for the same reasons. This really is mean to be a few helpers/types for dealing with callbacks and thread IPC mechanisms that are commonly used. async_xyz_cb and datatype are obvious and easy to add to imo.

Reply:
if `async` approach is better i don't mind deprecating `notify`. I seems that it is not that widely used (only onoff now? was suppose to be used by my old i2c async propsal).